llm_inference.py

```diff
@@ -420,6 +429,109 @@ def reshape_model_outputs(outputs: List[str], input_batch_size: int) -> List[Lis
+                    presence_penalty=self.args.presence_penalty,
+                    temperature=self.args.temperature,
+                    cohere_api_key=COHERE_API_KEY
+                )
+            else:
+                if self.args.model_name_or_path == 'openai-gpt-3.5-turbo':
+                    self.model = ChatOpenAI(
+                        model=model_name,
+                        temperature=self.args.temperature,
+                        max_tokens=self.args.max_new_tokens,
+                        top_p=self.args.top_p,
+                        frequency_penalty=self.args.frequency_penalty,
+                        presence_penalty=self.args.presence_penalty,
+                        openai_api_key=OPENAI_API_KEY
+                        )
+                else:
+                    # TODO: Consider adjusting parameter `n` (number of generations) or `best_of`
+                    self.model = OpenAI(
+                        model=model_name,
+                        temperature=self.args.temperature,
+                        max_tokens=self.args.max_new_tokens,
+                        top_p=self.args.top_p,
+                        frequency_penalty=self.args.frequency_penalty,
+                        presence_penalty=self.args.presence_penalty,
+                        openai_api_key=OPENAI_API_KEY
+                    )
+
+        if self.args.batch_size > 1:
+            warnings.warn("Batch size set to value >1. API models only support batch size of 1. "
+                          "Will change batch size to 1 automatically.")
+            self.args.batch_size = 1
+        # track the estimated cost of the API calls
+        self.estimated_cost = 0
+
+    def generate_from_model(self, inputs: List[str]) -> List[List[str]]:
+        """
+        queries the generation model for a given batch of inputs
+
+        Returns a list of lists of generated outputs (i.e. batch_size x num_return_sequences)
+        """
+
+        from langchain.callbacks import get_openai_callback
+        # breakpoint()
+        start_time = time.time()
+        if self.args.model_name_or_path == 'openai-gpt-3.5-turbo':
+            # handle chat inputs differently
+            from langchain.prompts.chat import ChatPromptTemplate, HumanMessagePromptTemplate
+            from langchain import LLMChain
+
+            # FIXME: this is a adapted from the LangChain example, but should be improved for our use case
+            human_template="{text}"
+            human_message_prompt = HumanMessagePromptTemplate.from_template(human_template)
+            chat_prompt = ChatPromptTemplate.from_messages([human_message_prompt])
+            chain = LLMChain(llm=self.model, prompt=chat_prompt)
+            with get_openai_callback() as cb:
+                outputs = chain.run(inputs[0])
+                logger.info(f'{cb}'.replace("\n", ", "))
+                self.estimated_cost += cb.total_cost
+        else:
+            with get_openai_callback() as cb:
```

Comment:
This is bound to throw an error if you call an Cohere model, probably?

Reply:
Nice spotting! You're right. This functionality is only available with OpenAI models atm (https://python.langchain.com/en/latest/modules/models/llms/examples/token_usage_tracking.html#how-to-track-token-usage). It's now fixed and tested with both OpenAI and Cohere models.